### PR TITLE
feat: Get rid of old error system

### LINKED ISSUE #633: Stop accepting strings in `GuppyError`

### PR DESCRIPTION
**Pull request by mark-koch**
Closes #633

Also updates a few last errors that I overlooked in the previous refactors

### CODE REVIEW

File: guppylang/error.py [hunk 1]
```diff
@@ -79,7 +28,7 @@ class GuppyTypeInferenceError(GuppyError):
     """Special Guppy exception for type inference errors."""
 
 
-class MissingModuleError(GuppyError):
+class MissingModuleError(Exception):
```
Comment (ss2165) [inline guppylang/error.py:31]: not a GuppyError?

File: guppylang/prelude/_internal/checker.py [hunk 2]
```diff
@@ -183,13 +164,22 @@ def check(self, args: list[ast.expr], ty: Type) -> tuple[ast.expr, Subst]:
 class NewArrayChecker(CustomCallChecker):
     """Function call checker for the `array.__new__` function."""
 
+    @dataclass(frozen=True)
+    class InferenceError(Error):
+        title: ClassVar[str] = "Cannot infer type"
+        span_label: ClassVar[str] = "Cannot infer the the type of this array"
```
Comment (ss2165) [inline guppylang/prelude/_internal/checker.py:170]:
```suggestion
        span_label: ClassVar[str] = "Cannot infer the type of this array"
```

File: guppylang/error.py (hunk 1 above)
Comment (mark-koch) [reply guppylang/error.py:31]: Yes, `GuppyError` is now only for stuff that goes wrong during compilation. This `MissingModuleError` is thrown immediately when you try to do `guppy.load()` with something that is not a module

File: tests/error/poly_errors/arg_mismatch4.err [hunk 3]
```diff
@@ -1,6 +1,8 @@
-Guppy compilation failed. Error in file $FILE:9
+Error: Expected expression of type `nat` (at $FILE:9:12)
+  | 
+7 | 
+8 | @guppy(module)
+9 | def main(x: array[int, bool]) -> None:
+  |             ^^^^^^^^^^^^^^^^ Expected expression of type `nat`, got type `bool`
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply tests/error/poly_errors/arg_mismatch4.err:6]: Yeah, again needs a refactor in type parsing to keep track of the span for individual arguments. See #637

**Review by ss2165 — APPROVED** [review]
👍

**Pull request merged**